File: src/common/web/services/ComponentService.ts
import { SemVer } from "semver";
import { ComponentInformationEvent } from "../api/component/ComponentEvents";
import { API_PROTOCOL_VERSION, checkProtocolCompatibility, ProtocolCompatibility } from "../api/Version";
import { WebComponent } from "../component/WebComponent";
import logging from "../core/logging/Logging";
import { Channel } from "../core/messaging/Channel";
import { ComponentState, useComponentStore } from "../data/stores/ComponentStore";
import { ConnectionState, useNetworkStore } from "../data/stores/NetworkStore";
import { Service } from "./Service";
import { ServiceContext } from "./ServiceContext";

/**
 * Creates the component service that reacts to basic messages.
 *
 * @param comp - The main component instance.
 *
 * @returns - The newly created service.
 */
export default function (comp: WebComponent): Service {
    return comp.createService("Component service", (svc: Service) => {
        const nwStore = useNetworkStore();
        const compStore = useComponentStore();

        svc.messageHandler(ComponentInformationEvent, (msg: ComponentInformationEvent, ctx: ServiceContext) => {
            // Notify of mismatching API protocol versions, which might lead to errors in network communication
            const protocolCompat = checkProtocolCompatibility(new SemVer(msg.api_protocol));
            if (protocolCompat == ProtocolCompatibility.NotCompatible) {
                logging.error("API major version mismatch; the affected components will not work together properly", "network", {
                    component: msg.comp_id,
                    got: msg.api_protocol,
                    want: API_PROTOCOL_VERSION.toString()
                });
            } else if (protocolCompat == ProtocolCompatibility.MaybeCompatible) {
                logging.warning("API minor version mismatch; the affected components might not work together properly", "network", {
                    component: msg.comp_id,
                    got: msg.api_protocol,
                    want: API_PROTOCOL_VERSION.toString()
                });
            }

            // This message is always received from the server side; we need to send our information in return
            const data = WebComponent.instance.data;
            ComponentInformationEvent.build(ctx.messageBuilder, data.compID, data.name, data.version.toString(), msg).emit(Channel.direct(msg.comp_id));

            // Our connection to the server is now ready to be used; save the remote info and change our internal state to 'Running'
            nwStore.serverInfo = msg.componentInformation();
            nwStore.serverChannel = Channel.direct(msg.comp_id);

            nwStore.connectionState = ConnectionState.Ready;
            compStore.componentState = ComponentState.Running;
        });
    });
}

File: src/common/py/services/component_service.py
from semantic_version import Version

from .client_service_context import ClientServiceContext
from .service import Service, ServiceContext
from ..api import (
    API_PROTOCOL_VERSION,
    check_protocol_compatibility,
    ProtocolCompatibility,
)
from ..component import BackendComponent
from ..core import logging


def create_component_service(comp: BackendComponent) -> Service:
    """
    Creates the component service that handles various basic messaging tasks.

    Args:
        comp: The main component instance.

    Returns:
        The newly created service.
    """
    from ..core.messaging import Channel
    from ..api.component import ComponentInformationEvent, ComponentProcessEvent

    svc = comp.create_service("Component service")

    @svc.message_handler(ComponentInformationEvent, is_async=True)
    def component_information(
        msg: ComponentInformationEvent, ctx: ServiceContext
    ) -> None:
        # Notify of mismatching API protocol versions, which might lead to errors in network communication
        compat = check_protocol_compatibility(Version(msg.api_protocol))
        if compat == ProtocolCompatibility.NOT_COMPATIBLE:
            logging.error(
                "API major version mismatch; the affected components will not work together properly",
                scope="network",
                component=msg.comp_id,
                got=msg.api_protocol,
                want=str(API_PROTOCOL_VERSION),
            )
        elif compat == ProtocolCompatibility.MAYBE_COMPATIBLE:
            logging.warning(
                "API minor version mismatch; the affected components might not work together properly",
                scope="network",
                component=msg.comp_id,
                got=msg.api_protocol,
                want=str(API_PROTOCOL_VERSION),
            )

        # If this message is received through the client, we need to send our information in return to the server; we also store the channel of the server for client components
        if ctx.is_entrypoint_client:
            remote_channel = Channel.direct(msg.comp_id)
            ClientServiceContext.set_remote_channel(remote_channel)

            data = BackendComponent.instance().data
            ComponentInformationEvent.build(
                ctx.message_builder,
                comp_id=data.comp_id,
                comp_name=data.name,
                comp_version=str(data.version),
                chain=msg,
            ).emit(remote_channel)

    @svc.message_handler(ComponentProcessEvent, is_async=True)
    def component_process(msg: ComponentProcessEvent, ctx: ServiceContext) -> None:
        # Listen to this event to avoid complains about unhandled messages
        pass

    return svc
